SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,528 @@
+/*
+ * Emitter.cs
+ * SnowplowTracker
+ * 
+ * Copyright (c) 2015 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ * 
+ * Authors: Joshua Beemster
+ * Copyright: Copyright (c) 2015 Snowplow Analytics Ltd
+ * License: Apache License Version 2.0
+ */
+
+using System;
+using System.Net;
+using System.Web;
+using System.IO;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Threading;
+using SnowplowTracker.Payloads;
+using SnowplowTracker.Enums;
+using SnowplowTracker.Requests;
+using SnowplowTracker.Storage;
+using SnowplowTracker.Collections;
+using UnityHTTP;
+
+namespace SnowplowTracker {
+	public class Emitter {
+
+		public bool synchronous = false;
+
+		private int POST_WRAPPER_BYTES = 88; // "schema":"iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-3","data":[]
+		private int POST_STM_BYTES = 22;     // "stm":"1443452851000",
+		private int FAIL_INTERVAL  = 10000;  // If all events failed to send
+
+		// Emitter loop variables
+		private readonly object emitLock = new object ();
+		private volatile bool sending = false;
+		private Thread emitThread;
+
+		// Emitter event store add variables
+		private ConcurrentQueue<TrackerPayload> payloadQueue = new ConcurrentQueue<TrackerPayload>();
+		private volatile bool consuming = false;
+		private Thread payloadConsumer;
+
+		// Emitter variables
+		private string endpoint;
+		private string collectorUri;
+		private HttpProtocol httpProtocol;
+		private HttpMethod httpMethod;
+		private int sendLimit;
+		private int bufferSize;
+		private long byteLimitGet;
+		private long byteLimitPost;
+		private EventStore eventStore;
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="SnowplowTracker.Emitter"/> class.
+		/// </summary>
+		/// <param name="endpoint">The collector endpoint uri</param>
+		/// <param name="protocol">What protocol to send under</param>
+		/// <param name="method">What method of sending to use</param>
+		/// <param name="sendLimit">The amount of events to pull from the database per sending attempt</param>
+		/// <param name="bufferSize">The maximum amount of events to put in a POST</param>
```

Comment:
Well it doesn't send every event, only a specific range at a time. I digress.  I think the occasional small events outweigh having to coordinate a timeout for that as well.  If a lot of events are getting added then we have big batches for that duration still while always staying on top of the amount of events in the database.

I have also been thinking regarding POST's.  If we have a byte-limit in place should we just remove the buffer size entirely and build POST's up until the limit to reduce the requests that way?